### ISSUE TITLE
new rule: max-bindings

### ISSUE THREAD

**Issue by gilligan** [eslint/eslint#5219]
#### motivation
There is already a rule that limits the maximum number of statements (`max-statements`). This new rule would instead limit the amount of const/let bindings per function.

#### type of rule
Like `max-statements` this would belong to the `stylistic` category

#### examples

The following would not create a warning:
```js
/* eslint max-bindings: [2, 3] */
const f = () => {
  const a = 1;
  const b = 2;
  return a + b;
};
```

The following would create a warning:
```js
/* eslint max-bindings: [2, 3] */
const f = () => {
  const a = 1;
  const b = 2;
  let c = 3;
  let d = 4;
  return a + b;
};
// => error This function has too many bindings (4). Maximum allowed is 3.
```
#### inclusion
Again referring to `max-statements` I think it would make sense to also include this rule as part of the eslint core rules.

**Comment by platinumazure** [comment]
Should `var` statements be included? And should the let/const bindings only be counted at the block level, or function?

Example of what I mean by my last question:

```js
/* eslint max-bindings: [2, 3] */
const f = (foo) => {
    let a, b;
    if (foo) {
        let c, d;
    }
};
```

Four total bindings in the function, two in each block scope. Should the rule warn or not? And if they were var declarations, should this rule warn in that case?

**Comment by michaelficarra** [comment]
@platinumazure Why would var declarations be counted differently? That would make no sense.

@gilligan Would parameter bindings and catch bindings be included in the count?

**Comment by platinumazure** [comment]
@michaelficarra I asked because the original proposal didn't include var
bindings at all, so there was ambiguity. Maybe there's another rule that
covers var bindings already. Maybe the user meant for them to be included,
as you suggest. Maybe the user didn't think about it at all. Don't make
baseless assumptions about why a question was asked just because you think
you know the answer.

**Comment by michaelficarra** [comment]
Well then let's confirm everything to be on the safe side. @gilligan Are function declaration bindings counted? How about generator function declaration bindings? Are class declaration bindings counted? Are function expression name bindings counted? How about generator function expression name bindings? Are class declaration immutable inner bindings counted? Are class expression name bindings counted?

**Comment by ilyavolodin** [comment]
@gilligan Could you please expend on what this rule is trying to solve? There's a pretty good understand (at least I think there is) of why too many function arguments are a bad thing, but I'm not sure I understand why too many local variables are bad.

**Comment by gilligan** [comment]
Wow, quite some responses here. Thanks for your interest. I clearly was not as well prepared as I had thought.

@ilyavolodin I think in the end both are about keeping functions small and adhering to the single responsibility principle and readability.

```js
function f (a, b) {
 const v1 = t1(a);
 const v2 = t2(b);
 const v3 = t3(a, v1);
 const v4 = t4(v2, v3);
...
return vn;
};
```

Sure enough the bindings `v1`..`vn` are all local to the function scope of `f` but the human brain is pretty bad at juggling around more than three or four values at once. This leads to `wait a second, what is v3 here, how did that end up there.. *trace back* .. ah it went from the command line argument then was transformed by ...`

To answer the other questions: 
* scope: function scope
* bindings: `const`, `let`, `var` statements

**Comment by ilyavolodin** [comment]
@gilligan Thanks for the answer!
@eslint/eslint-team Anyone wants to champion a rule like this?

**Comment by platinumazure** [comment]
Not in the core team, but I'm wondering if this isn't perhaps covered by `max-statements`. I'd like to hear from @gilligan on what this new rule would buy us that `max-statements` could not.

**Comment by lo1tuma** [comment]
@platinumazure e.g. `var foo, bar, baz;` would be counted as one statement, but it creates 3 bindings.

**Comment by platinumazure** [comment]
I understand that. My question is, are we sure the variable declarations
represent the cognitive load, or would the statements that initialize and
use the variables represent the true cognitive load?

**Comment by BYK** [comment]
I think a rule like this would be useful and I think (re @platinumazure) the bindings are just another metric of complexity in addition to number of statements.

I'm willing to give a hand to move the rule forward (either via code or supporting @gilligan if he's willing to create a PR).

**Comment by ilyavolodin** [comment]
Looks like @BYK is championing this rule, so marking as accepted.

**Comment by nzakas** [comment]
To move forward, I think the following needs to be addressed:

1. Is "bindings" the correct term for this rule?
1. Can this really be done just be considering function scope? For instance, what about this:

```js
function x() {
    const y = "a";

    if (foo) {
        let z = "b";
    }
}
```

Would you consider `z` to be part of the overall bindings for the function? Would it count towards the max even though it ceases to exist once the program flows out of the `if` statement?

I think this still needs a fair bit of clarification before an implementation can begin.

**Comment by gilligan** [comment]
@nzakas fair enough, I agree that first of all the rule name needs refinement. I will provide some suggestions tomorrow and also try to address the scope question.

**Comment by nzakas** [comment]
Sounds good. For the record, I'm concerned about this rule as I am unclear as to why limiting the number of variables is a good idea. In my experience, I've generally encouraged people to use more variables rather than fewer, as I've found it aids clarity and decreases cognitive load (like `no-magic-numbers`).

**Comment by BYK** [comment]
May be this is better as a plugin then?

**Comment by nzakas** [comment]
Yeah, I'm thinking this is best as a custom rule. 

I'm the future, let's make sure to discuss rule details a bit before championing them for core.

@gilligan you can still create this rule for yourself. If you do, feel free to comment back here to share.
